build: detect and append warning flags only if they are available

fanquake opened the pull request:
This is a followup to #1046. Opening a draft for feedback.

Rather than appending flags directly, or testing for various old versions of GCC, check if the flag will work with the compiler, and append it to CFLAGS if so. `-Werror` is used, so that flags which aren't understood, i.e Clang and `-Wlogical-op`, don't end up appended (AX_CHECK_COMPILE_FLAG ignores warnings).

This also removes code which is commented out, and moves the automake options into configure.

Note: Why is `-Wstrict-aliasing` explicitly turned on (it's also part of `-Wall`), when strict aliasing is disabled using `-fno-strict-aliasing`? Looking at the git history is seems that it was added before strict aliasing was disabled.

azat commented:
>Rather than appending flags directly, or testing for various old versions of GCC, check if the flag will work with the compiler

Great!

>Note: Why is -Wstrict-aliasing explicitly turned on (it's also part of -Wall), when strict aliasing is disabled using -fno-strict-aliasing? Looking at the git history is seems that it was added before strict aliasing was disabled.

Yes, no point in `-Wstrict-aliasing`

fanquake commented:
I've rebased on master and marked as ready for review.

> Yes, no point in -Wstrict-aliasing

Ok. I've added a commit to drop `-Wstrict-aliasing`.

> LGTM, how about making the same in cmake then?

I don't have much experience with CMake, and don't currently use it anywhere that I'm compiling libevent, so it might be better if someone else made the equivalent changes. However, if that would be a blocker for this to be merged I can take a look.

azat commented:
>I don't have much experience with CMake, and don't currently use it anywhere that I'm compiling libevent, so it might be better if someone else made the equivalent changes. However, if that would be a blocker for this to be merged I can take a look.

Well, cmake and autotools should be in sync, so yes this is a blocker.
But no worries, I can take a look at cmake (I will do this tomorrow, so you will have time to take a look if you want to)

azat commented:
>But no worries, I can take a look at cmake (I will do this tomorrow, so you will have time to take a look if you want to)

No need to, cmake already uses `check_c_compiler_flag`
